compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -363,20 +366,55 @@ fn loose_check<'cx, 'tcx>(
     selcx: &mut SelectionContext<'cx, 'tcx>,
     o: &PredicateObligation<'tcx>,
 ) -> bool {
+    // FIXME: the call to `selcx.predicate_may_hold_fatal` below should be ported
+    // to the canonical trait query form, `infcx.predicate_may_hold`, once
+    // the new system supports intercrate mode (which coherence needs).
     !selcx.predicate_may_hold_fatal(o)
 }
 
 fn negative_impl_exists<'cx, 'tcx>(
     selcx: &SelectionContext<'cx, 'tcx>,
+    region_context: DefId,
     o: &PredicateObligation<'tcx>,
 ) -> bool {
-    let infcx = selcx.infcx();
+    let infcx = selcx.infcx().clone();
     let tcx = infcx.tcx;
     o.flip_polarity(tcx)
-        .as_ref()
         .map(|o| {
-            // FIXME This isn't quite correct, regions should be included
-            selcx.infcx().predicate_must_hold_modulo_regions(o)
+            let mut fulfillment_cx = FulfillmentContext::new();
+            fulfillment_cx.register_predicate_obligation(infcx, o);
+
+            let errors = fulfillment_cx.select_all_or_error(infcx);
+            if !errors.is_empty() {
+                return false;
+            }
+
+            // let body_id_map: FxHashMap<_, _> = infcx
+            //     .inner
+            //     .borrow()
+            //     .region_obligations()
+            //     .iter()
+            //     .map(|&(id, _)| (id, vec![]))
+            //     .collect();
+            let param_env = ty::ParamEnv::empty();
+
+            // infcx.process_registered_region_obligations(
```

Comment:
Try this (similar to [here](https://github.com/Mark-Simulacrum/rust/blob/66dfaa16409868e12ef2168f39516ae982b1fe1f/compiler/rustc_typeck/src/check/regionck.rs#L376-L388)):

```rust
let outlives_environment = OutlivesEnvironment::new(impl1_env);
self.infcx.process_registered_region_obligations(
    outlives_environment.region_bound_pairs_map(),
    Some(self.tcx.lifetimes.re_root_empty),
    impl1_env
);
```